### PR TITLE
Update BinaryFormat with details about encoding

### PR DESCRIPTION
**Pull request by titzer**
Unfortunately @kg is unable to continue with #520 which updates BinaryEncoding.md to be more in line with the V8 format embodied in the design doc. Unfortunately the branch from which she made the pull request is in her private repo, and I could not edit it. So I started with her changes and made this pull request to continue the discussion.

### CODE REVIEW

File: BinaryEncoding.md [hunk 1]
```diff
@@ -1,122 +1,398 @@
...
+The binary encoding is a dense representation of module information that enables 
+small files, fast decoding, and reduced memory usage.
+See the [rationale document](Rationale.md#why-a-binary-encoding) for more detail.
+
+The encoding is split into three layers:
+
+* **Layer 0** is a simple pre-order encoding of the AST and related data structures.
+  The encoding is dense and trivial to interact with, making it suitable for
+  scenarios like JIT, instrumentation tools, and debugging.
+* **Layer 1** provides structural compression on top of layer 0, exploiting
```
Comment (kripken) [inline BinaryEncoding.md]: same question as in previous pull - this change removes the text "this is not meant to be standardized currently, but we'll evaluate later". was that intentional? i.e., are you suggesting that the MVP include layer 1?

Comment (titzer) [inline BinaryEncoding.md]: I've added text to the paragraph below that makes it more clear that we won't standardize Layers 1 and 2 until after experimentation.

File: BinaryEncoding.md [hunk 2]
```diff
@@ -1,122 +1,400 @@
...
+occur incrementally. For example, Layer 1 and Layer 2 encoding techniques can be
+experimented with by application-level decompressing to the layer below. As 
+compression techniques  stabilize, they can be standardized and moved into native 
+implementations.
+
+# Data types
+
+### uint8
+A single-byte unsigned integer.
+
```
Comment (JSStats) [inline BinaryEncoding.md]: Is a signed int8 data type needed for the i8.const operator immediate argument?

The f32 and f64 immediate data types.

File: BinaryEncoding.md [hunk 3]
```diff
@@ -1,122 +1,400 @@
...
-      or affect anything outside of the initialization of the index spaces; decoders would
-      remain versionless and simply add cases for new *names* (as with current JavaScript parsers).
-
-## Proposals
-
-The native prototype built for [V8](https://github.com/WebAssembly/v8-native-prototype)
-implements a binary format that embodies most, but not all of the ideas in this document.
-It is described in detail in a [public design doc](https://docs.google.com/a/google.com/document/d/1761v1AfhFM5kE8NArF_PyXcl-iVh0Dx3InOrmcyIoiI/pub) and a [copy of the original](https://docs.google.com/document/d/1-G11CnMA0My20KI9D7dBR6ZCPOBCRD0oCH6SHCPFGx0/edit?usp=sharing).
+
+  * Given a call AST node: `struct Call { uint32_t callee_index; vector<AstNode*> args; }`
```
Comment (JSStats) [inline BinaryEncoding.md]: The use of a C style struct seems out of place and unnecessary.

File: BinaryEncoding.md [hunk 4]
```diff
@@ -1,122 +1,400 @@
...
+    * First write the binary sequence of `Call` (uint8)
```
Comment (JSStats) [inline BinaryEncoding.md]: 'binary sequence of `Call`' is not very meaningful. The word 'opcode' seems appropriate, and the data type uint8.

File: BinaryEncoding.md [hunk 5]
```diff
@@ -1,122 +1,400 @@
...
+| ----- |  ----- | ----- | 
+| id = `0x01` | `uint8` | section identifier for signatures |
+| count | `varuint32` | count of signature entries to follow | 
+| entries | `signature_entry*` | repeated signature entries as described below |
+
+#### Signature entry
+| Field | Type | Description |
+| ----- |  ----- | ----- | 
+| param_count | `uint8` | the number of parameters to the function |
+| return_type | `local_type?` | the return type of the function, with `0` indicating no return type |
```
Comment (JSStats) [inline BinaryEncoding.md]: Could define the local_type above.

File: BinaryEncoding.md [hunk 6]
```diff
@@ -1,122 +1,400 @@
...
+| param_types | `local_type*` | the parameter types of the function, with<br>`1` indicating type `i32`<br>`2` indicating type `i64`<br>`3` indicating type `f32`<br>`4` indicating type `f64` |
```
Comment (JSStats) [inline BinaryEncoding.md]: Oh, perhaps put these local_types in a separate table.

**Comment by titzer** [comment]
@JSStats see https://github.com/WebAssembly/design/commit/fd8db871fc5d2a89b0e6aa2041695f5e6f655d00 for definition of value_type.

File: BinaryEncoding.md [hunk 7]
```diff
@@ -1,122 +1,400 @@
...
+| flags | `uint8` | always | flags indicating attributes of a function <br>bit `0` : a name is present<br>bit `1` : the function is an import<br>bit `2` : the function has local variables<br>bit `3` : the function is an export |
+| signature | `uint16` | always | index into the Signature section |
+| name | `uint32` | `flags[0] == 1` | name of the function as an offset within the module |
+| i32 count | `uint16` | `flags[2] == 1` | number of i32 local variables |
+| i64 count | `uint16` | `flags[2] == 1` | number of i64 local variables |
+| f32 count | `uint16` | `flags[2] == 1` | number of f32 local variables |
+| f64 count | `uint16` | `flags[2] == 1` | number of f64 local variables |
+| body size | `uint16` | `flags[0] == 0` | size of function body to follow, in bytes |
+| body | `bytes` | `flags[0] == 0` | function body |
+
```
Comment (JSStats) [inline BinaryEncoding.md]: Is there a constraint that a name is required for an import or export? Is a name allowed is neither?

The body_size and body depend on bit 1 (not an import)?

Comment (titzer) [inline BinaryEncoding.md]: Yes, imported and exported functions must have names.

File: BinaryEncoding.md [hunk 8]
```diff
@@ -1,122 +1,400 @@
...
+| ----- |  ----- | ----- |
+| id = `0x11` | `uint8` | section identifier for globals |
+| size | `varuint32` | size of this section in bytes | 
+| body | `bytes` | contents of this section |
+
+# AST Encoding
+
+Function bodies consist of a dense pre-order encoding of an [Abstract Syntax Tree](AstSemantics.md).
+Each node in the abstract syntax tree corresponds to an operator, such as `i32.add` or `if` or `block`.
+Most operators encode as a single `uint8`, but some require *immediates* which immediately follow the
```
Comment (JSStats) [inline BinaryEncoding.md]: 'Most operators encode as a single `uint8' does not read well, there are few 'operators' that encoding in a single uint8. Perhaps: operators are encoded by a uint8 opcode followed by fixed immediate arguments if any and then their child nodes if any.

File: BinaryEncoding.md [hunk 9]
```diff
@@ -1,122 +1,400 @@
...
+
+The `tableswitch` operator has as complex immediate operand which is encoded as follows:
+
+| Field | Type | Description |
+| ---- | ---- | ---- |
+| case_count | `uint16` | number of cases in the case_table |
+| target_count | `uint16` | number of targets in the target_table |
+| target_table | `uint16*` | target entries where<br>`>= 0x8000` indicates an outer block to which to break<br>`<= 0x8000` indicates a case to which to jump |
+
+The table switch operator is then immediately followed by `case_count` case expressions which by default fall through to each other.
```
Comment (JSStats) [inline BinaryEncoding.md]: This does not read well. It seems to be followed by target_count-1 case uint16 values, then one default case uint16 value, then the key child expression, and then case_count child expressions.

File: BinaryEncoding.md [hunk 10]
```diff
@@ -1,122 +1,400 @@
...
+
+## Basic operators ([described here](AstSemantics.md#constants))
+| Name | Opcode | Immediate | Description |
+| ---- | ---- | ---- | ---- |
+| `i8.const` | `0x09` | value = `uint8` | a constant value, signed extended to type `i32`  |
```
Comment (JSStats) [inline BinaryEncoding.md]: uint8 -> int8

File: BinaryEncoding.md [hunk 11]
```diff
@@ -1,122 +1,400 @@
...
+| `i32.const` | `0x0a` | value = `uint32` | a constant value interpreted as `i32` |
+| `i64.const` | `0x0b` | value = `uint64` | a constant value interpreted as `i64` |
+| `f64.const` | `0x0c` | value = `uint64` | a constant value interpreted as `f64` |
+| `f32.const` | `0x0d` | value = `uint32` | a constant value interpreted as `f32` |
```
Comment (JSStats) [inline BinaryEncoding.md]: The 'constant value interpreted as' does not read well - the integers are really neither signed or unsigned just bit storage widths.

Seems fine to add data types for the float immediates.

File: BinaryEncoding.md [hunk 12]
```diff
@@ -1,122 +1,400 @@
...
+| `get_local` | `0x0e` | local_index = `varuint32` | read a local variable or parameter |
+| `set_local` | `0x0f` | local_index = `varuint32` | write a local variable or parameter |
```
Comment (JSStats) [inline BinaryEncoding.md]: Should also note that set_local has a child expression to read.

File: BinaryEncoding.md [hunk 13]
```diff
@@ -1,122 +1,400 @@
...
+| `load_global` | `0x10` | index = `varuint32` | * nonstandard internal opcode |
+| `store_global` | `0x11` | index = `varuint32` | * nonstandard internal opcode |
+| `call_function` | `0x12` | function_index = `varuint32` | call a function by its index |
+| `call_indirect` | `0x13` | signature_index = `varuint32` | call a function indirect with an expected signature |
```
Comment (JSStats) [inline BinaryEncoding.md]: call_function has one child expression for each argument from the function signature.

call_indirect has one child expression for the index, plus one for each argument from the signature.

File: BinaryEncoding.md [hunk 14]
```diff
@@ -1,122 +1,400 @@
...
+| `f32.store` | `0x35` | `memory_immediate` | store to memory |
+| `f64.store` | `0x36` | `memory_immediate` | store to memory |
+| `memory_size` | `0x3b` |  | query the size of memory |
+| `grow_memory` | `0x39` |  | grow the size of memory |
+
+The `memory_immediate` type is encoded as follows:
+
+| Name | Type | Present? | Description |
+| ---- | ---- | ---- | ---- |
+| flags | `uint8` | always | a bitfield where<br>bit `4` indicates an offset follows<br>bit `7` indicates natural alignment |
```
Comment (JSStats) [inline BinaryEncoding.md]: Could define the meaning of 0 and 1 for these flags.

File: BinaryEncoding.md [hunk 15]
```diff
@@ -1,122 +1,400 @@
...
+| offset | `varuint32` | `flags[4] == 1` | the value of the offset |
+
+## Simple operators ([described here](AstSemantics#32-bit-integer-operators))
+
+| Name | Opcode | Immediate | Description |
+| ---- | ---- | ---- | ---- |
+| `i32.add` | `0x40` | | |
+| `i32.sub` | `0x41` | | |
```
Comment (JSStats) [inline BinaryEncoding.md]: Should at least define the number of arguments for each here so the child nodes can be read.

File: BinaryEncoding.md [hunk 16]
```diff
@@ -1,122 +1,400 @@
...
+| `f32.convertf64` | `0xac` | | |
+| `f32.reinterpreti32` | `0xad` | | |
+| `f64.sconverti32` | `0xae` | | |
+| `f64.uconverti32` | `0xaf` | | |
+| `f64.sconverti64` | `0xb0` | | |
+| `f64.uconverti64` | `0xb1` | | |
+| `f64.convertf32` | `0xb2` | | |
+| `f64.reinterpreti64` | `0xb3` | | |
+| `i32.reinterpretf32` | `0xb4` | | |
+| `i64.reinterpretf64` | `0xb5` | | |
```
Comment (JSStats) [inline BinaryEncoding.md]: I think you should use the groups names for all these rather than the v8 names. There has been a lot of confusion in this area. Here's my table:
```
    (i32.trunc_s/f32     #x9d  (f32) nil              i32) ; i32.sconvert_f32
    (i32.trunc_s/f64     #x9e  (f64) nil              i32) ; i32.sconvert_f64
    (i32.trunc_u/f32     #x9f  (f32) nil              i32) ; i32.uconvert_f32
    (i32.trunc_u/f64     #xa0  (f64) nil              i32) ; i32.uconvert_f64
    (i32.wrap/i64        #xa1  (i64) nil              i32) ; i32.convert.i64
    (i64.trunc_s/f32     #xa2  (f32) nil              i64) ; i64.sconvert_f32
    (i64.trunc_s/f64     #xa3  (f64) nil              i64) ; i64.sconvert_f64
    (i64.trunc_u/f32     #xa4  (f32) nil              i64) ; i64.uconvert_f32
    (i64.trunc_u/f64     #xa5  (f64) nil              i64) ; i64.uconvert_f64
    (i64.extend_s/i32    #xa6  (i32) nil              i64) ; i64.sconvert_i32
    (i64.extend_u/i32    #xa7  (i32) nil              i64) ; i64.uconvert_i32
    (f32.convert_s/i32   #xa8  (i32) nil              f32) ; f32.sconvert_i32
    (f32.convert_u/i32   #xa9  (i32) nil              f32) ; f32.uconvert_i32
    (f32.convert_s/i64   #xaa  (i64) nil              f32) ; f32.sconvert_i64
    (f32.convert_u/i64   #xab  (i64) nil              f32) ; f32.uconvert_i64
    (f32.demote/f64      #xac  (f64) nil              f32) ; f32.convert_f64
    (f32.reinterpret/i32 #xad  (i32) nil              f32) ; f32.reinterpret_i32
    (f64.convert_s/i32   #xae  (i32) nil              f64) ; f64.sconvert_i32
    (f64.convert_u/i32   #xaf  (i32) nil              f64) ; f64.uconvert_i32
    (f64.convert_s/i64   #xb0  (i64) nil              f64) ; f64.sconvert_i64
    (f64.convert_u/i64   #xb1  (i64) nil              f64) ; f64.uconvert_i64
    (f64.promote/f32     #xb2  (f32) nil              f64) ; f64.convert_f32
    (f64.reinterpret/i64 #xb3  (i64) nil              f64) ; f64.reinterpret_i64
    (i32.reinterpret/f32 #xb4  (f32) nil              i32) ; i32.reinterpret_f32
    (i64.reinterpret/f64 #xb5  (f64) nil              i64) ; i64.reinterpret_f64
```

File: BinaryEncoding.md [hunk 17]
```diff
@@ -1,122 +1,408 @@
...
+| Field | Type | Description |
+| ----- |  ----- | ----- |
+| id = `0x11` | `uint8` | section identifier for globals |
+| size | `varuint32` | size of this section in bytes | 
+| body | `bytes` | contents of this section |
+
+# AST Encoding
+
+Function bodies consist of a dense pre-order encoding of an [Abstract Syntax Tree](AstSemantics.md).
+Each node in the abstract syntax tree corresponds to an operator, such as `i32.add` or `if` or `block`.
```
Comment (JSStats) [inline BinaryEncoding.md]: Perhaps at this level the spec could focus on decoding the AST into the operators with their arguments and child nodes, and could defer their checking and interpretation to following sections. If so then `block` might be defined here as 'a sequence of 'count' child expressions. Similarly for `loop`. `Return` would be defined as having a child to read if the function has results. `if` defined as having two child expressions. `if_else` as having three child expressions. `select`' as having three child expressions. `br` - two. `br_if` - three. This would at least get the AST decoded into an internal form.

File: BinaryEncoding.md [hunk 18]
```diff
@@ -1,122 +1,408 @@
...
+  tree nodes.
+* **Layer 2** applies generic compression techniques, already available
+  in browsers and other tooling. Algorithms as simple as gzip can deliver
+  good results, but more sophisticated algorithms like 
+  [LZHAM](https://github.com/richgel999/lzham_codec) and
+  [Brotli](https://datatracker.ietf.org/doc/draft-alakuijala-brotli/) are able
+  to deliver dramatically smaller files.
+
+Most importantly, the layering approach allows development and standardization to
+occur incrementally. For example, Layer 1 and Layer 2 encoding techniques can be
```
Comment (lukewagner) [inline BinaryEncoding.md]: Layer 2 is already in the browser (via HTTP `Content-Encoding` `gzip`/`bro`).  Hopefully one day it would also be provided as a Streams transformer, so probably this isn't something we're even considering.

File: BinaryEncoding.md [hunk 19]
```diff
@@ -1,122 +1,408 @@
...
+  specific knowledge about the nature of the syntax tree and its nodes.
+  The structural compression introduces more efficient encoding of values, 
+  rearranges values within the module, and prunes structurally identical
+  tree nodes.
+* **Layer 2** applies generic compression techniques, already available
+  in browsers and other tooling. Algorithms as simple as gzip can deliver
+  good results, but more sophisticated algorithms like 
+  [LZHAM](https://github.com/richgel999/lzham_codec) and
+  [Brotli](https://datatracker.ietf.org/doc/draft-alakuijala-brotli/) are able
+  to deliver dramatically smaller files.
```
Comment (lukewagner) [inline BinaryEncoding.md]: Brotli is now in multiple browsers and is pretty competitive with LZHAM.  Maybe reword to just "Layer 2 applies generic compression algorithms, like [gzip] and [brotli], that are already available in browsers and other tooling."

File: BinaryEncoding.md [hunk 20]
```diff
@@ -1,122 +1,408 @@
...
+### End section
+This indicates the end of the sections. Additional data that follows this section marker is not interpreted
+by the decoder. It can used, for example, to store function names or data segment bodies.
+
+| Field | Type | Description |
+| ----- |  ----- | ----- | 
+| id = `0x06` | `uint8` | section identifier for end of sections |
+
+
+### Nonstandard: Globals section
```
Comment (lukewagner) [inline BinaryEncoding.md]: I'd remove this section since it's, well, non-standard :)

File: BinaryEncoding.md [hunk 21]
```diff
@@ -1,122 +1,408 @@
...
+A global variable entry describes a variable outside the WASM linear memory. A Global variable can only be loaded and
+stored to from dedicated instructions.
+
+| Field | Type | Description |
+| ----- |  ----- | ----- | 
+| name | `uint32` | name of the global variable as an offset to a string in the module | 
+| type | `uint8` | the type of the global, as a memory type |
+| exported | `uint8` | a boolean indicating whether the global variable is exported |
+
+### WorkInProgress: WLL section
```
Comment (lukewagner) [inline BinaryEncoding.md]: If we can switch to string section names then there shouldn't be a need to "reserve" section identifiers (the power of strings!) like this so I'd remove this section since it's also nonstandard.

File: BinaryEncoding.md (hunk 19 above)
Comment (JSStats) [inline BinaryEncoding.md]: The phrase 'applies generic compression techniques' seems erroneous too. Brotli has a very simple predictor that has no knowledge of the AST and surely it will be possible to do better with a better predictor. It's not clear that layer-1 can transform the AST to fit the Brotli predictor, it might be possible, but it still seems an open issue. Anyway, this PR seems a good step forward, and surely such points can be clarified in follow up PRs.

File: BinaryEncoding.md (hunk 21 above)
Comment (JSStats) [inline BinaryEncoding.md]: @lukewagner The WLL is being developed by a group external to the wasm community group and you have absolutely no standing to declare it non-standard, please stop making these claims.

Naming sections with practical strings would not solve the problem because it still does not manage the name space and I believe it is as ridiculous as suggesting that other web standards should be developed without management of name spaces. I guess the WLL could encode a signature etc in the string to try and avoid name clashes, but will a group championing accessibility of wasm really have to battle yourself and Mozilla?

CC: Chris Beard CEO Mozilla Corp.

Comment (lukewagner) [inline BinaryEncoding.md]: @JSStats It's literally not part of the wasm standard.  It shouldn't be hard to choose a string that has a low likelihood of collision.  At some later time it could become standardized based on popularity/usage.  Save the theatrics.

Comment (JSStats) [inline BinaryEncoding.md]: @lukewagner This PR includes recognition of the WLL section in the wasm draft standard. Google have included it. You are opposing this. I have taken up the matter of Mozilla's representation on this issue with Mozilla's management, and I did just write to them again. I do not engage in 'theatrics'.

With my WebAssembly group member hat on I oppose the group having a position that the wider web community just 'choose a string that has a low likelihood of collision', and I believe we should acknowledge other uses.

If the WLL section is poking people in the eye then I would be just as happy for this PR to be rewritten so that all other sections follow the same conventions of having a declared size so they can be skipped and with a table of codes/strings acknowledged uses that for now would just include the WLL use.  This seems a very small burden for the group and a constructive thing to do, but it the group really does not want any burden or responsibility for managing the name space then I guess it can be done elsewhere.

Comment (jfbastien) [inline BinaryEncoding.md]: Your opposition is noted, but if you want to be taken seriously I highly recommend stopping with the theatrics such as trying to call our respective bosses / CEOs, and positioning your opinion as that of the wider web community. This isn't how productive discourse works: witness our (Goog/Moz) frequent disagreements about WebAssembly which get resolved amicably. You've been very agreeable in January, but the last few days haven't been so. Please reconsider.

Comment (JSStats) [inline BinaryEncoding.md]: Dear Chris Beard, please see the personal comments form a Chair of this community group @jfbastien who while not your responsibility I believe should have been censured by your Mozilla representatives. Your representative Luke Wagner holds a position of Chair in this group.

@jfbastien Your depiction of my comments as  'theatrics' is defamatory. Your attempts to stop me communicating with the organisations being represented here is nothing short of bullying. I have made no claim that my opinions are that 'of the wider web community' and suggesting that I have and using it to demonize me is dishonest controlling behaviour. You do not have my confidence as Chair, and I ask you to please step aside.

Comment (titzer) [inline BinaryEncoding.md]: @JSStats 

Let's keep this discourse polite and treat each other with respect. CC'ing or threatening to CC the CEO of a corporation over a minor disagreement w.r.t. a binary format is neither going to be productive (our organizations do not work like that) nor is it acceptable behavior (it is aggressive and threatening). It saddens me to observe that despite your recent and welcome cooperativeness, untoward escalation and threats are appearing again. I personally accepted your patch to V8 to allow the WLL section because it was "no big deal" and included it on this PR along the vein of documenting the existing format. An open and dispassionate discussion about whether to include it is entirely expected and welcome. Threats are not.

Comment (JSStats) [inline BinaryEncoding.md]: @titzer Your comments are a gross violation of our code of conduction and you have significantly defamed me. I have every right to contact who I please regarding the business of this group. Members have no right to cry about other members undertaking the business of this group and to use this to attempt to restrict interactions and otherwise control people - this is simply inappropriate controlling behaviour and unprofessional behaviour. Would you please make a public retraction and apology.

File: BinaryEncoding.md [hunk 22]
```diff
@@ -1,122 +1,408 @@
...
+| `return` | `0x14` | | return zero or one value from this function |
+| `unreachable` | `0x15` | | trap immediately |
+
+The `tableswitch` operator has as complex immediate operand which is encoded as follows:
+
+| Field | Type | Description |
+| ---- | ---- | ---- |
+| case_count | `uint16` | number of cases in the case_table |
+| target_count | `uint16` | number of targets in the target_table |
+| target_table | `uint16*` | target entries where<br>`>= 0x8000` indicates an outer block to which to break<br>`<= 0x8000` indicates a case to which to jump |
```
Comment (mbebenita) [inline BinaryEncoding.md]: Should read ...  `< 0x8000` indicates a case to which to jump.

File: BinaryEncoding.md (hunk 8 above)
Comment (titzer) [inline BinaryEncoding.md]: Done.

File: BinaryEncoding.md (hunk 10 above)
Comment (titzer) [inline BinaryEncoding.md]: Done.

File: BinaryEncoding.md (hunk 12 above)
Comment (titzer) [inline BinaryEncoding.md]: The arity of operators (i.e. how many and what type child nodes should have) is described in AstSemantics.

File: BinaryEncoding.md (hunk 13 above)
Comment (titzer) [inline BinaryEncoding.md]: Same as above.

File: BinaryEncoding.md (hunk 15 above)
Comment (titzer) [inline BinaryEncoding.md]: That's just redundant with AstSemantics IMO.

File: BinaryEncoding.md (hunk 16 above)
Comment (titzer) [inline BinaryEncoding.md]: I just did a dump and we can rename them later.

File: BinaryEncoding.md (hunk 19 above)
Comment (titzer) [inline BinaryEncoding.md]: Done.

File: BinaryEncoding.md (hunk 21 above)
Comment (titzer) [inline BinaryEncoding.md]: I'll suggest we just leave the byte encodings in for now and hash out the details of strings later, as the point of this PR is to embody what exists now vs making major changes.

**Comment by titzer** [comment]
I believe I've addressed most of the comments here so far, so maybe we can push this PR past through the difficult discourse to a mergeable state?

**Comment by JSStats** [comment]
Concession https://codereview.chromium.org/1695233002/

**Comment by sunfishcode** [comment]
lgtm. There's obviously lots of work remaining to be done on the binary format, but this PR does describe the current prototype, and merging this PR will let us continue the work with one topic per PR instead of this one PR, which should make discussion easier.

**Comment by lukewagner** [comment]
Agreed; we can iterate further in separate PRs.

**Comment by jfbastien** [comment]
lgtm

**Comment by rossberg-chromium** [comment]
lgtm
